### ISSUE TITLE
Replicating Example for JOSS Submission

### ISSUE THREAD

**Issue by manuel-g-castro** [dtkerns/d2#5]
Hello, @dtkerns:

For the review of your submission to JOSS, openjournals/joss-reviews/issues/6111, I am trying to replicate the examples.

I have pulled the `ubuntu:latest` docker image to run the testing of your software. I had to change the default symlink `cc` to point to `/usr/bin/clang` (in this image it was pointing to `gcc`), otherwise both apps would not be compiled.

But, even with the llvm capable compiler set, running `d2` prints some errors due to files not existing (see log below).  Moreover, all the files produced by it are empty or really tiny. 

I am unsure that this is the desired outcome. 

```
root@6ce3f6fc9c55:/home/root/d2/output# find -type f -exec ls -l {} \;
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./app1/app1.sb
-rw-r--r-- 1 root root 12 Mar  2 23:20 ./app1/app1.dot
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./app1/app1.loop
-rw-r--r-- 1 root root 0 Mar  2 22:39 ./app1/app1.const
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./bbselected
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./bbsel_norm
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./lobb
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./bbdb
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./bbsel_norm.bb
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./loopdb
-rw-r--r-- 1 root root 30 Mar  2 23:20 ./loll
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./bbsbdb
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./app2/app2.loop
-rw-r--r-- 1 root root 12 Mar  2 23:20 ./app2/app2.dot
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./app2/app2.sb
-rw-r--r-- 1 root root 0 Mar  2 22:33 ./app2/app2.const
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./bbcdb
-rw-r--r-- 1 root root 32 Mar  2 23:20 ./lollg
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./sbselected
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./rankorder
-rw-r--r-- 1 root root 0 Mar  2 23:20 ./ronormfiles
```

```
root@6ce3f6fc9c55:/home/root/d2/example# d2 /home/root/d2/output/       
/home/root/d2/bin/d2: line 14: [: 14.0.0: integer expression expected
/home/root/d2/bin
/home/root/d2/output
for d in app1 app2; do \
	make -C $d clean; \
done
make[1]: Entering directory '/home/root/d2/example/app1'
rm -f app1 *.ll *.llg
make[1]: Leaving directory '/home/root/d2/example/app1'
make[1]: Entering directory '/home/root/d2/example/app2'
rm -f app2 *.ll *.llg
make[1]: Leaving directory '/home/root/d2/example/app2'

real	0m0.037s
user	0m0.021s
sys	0m0.015s
make -C app1 
make[1]: Entering directory '/home/root/d2/example/app1'
clang -std=c99 -S -emit-llvm app1.c -o app1.ll
make[1]: Leaving directory '/home/root/d2/example/app1'
make -C app2 
make[1]: Entering directory '/home/root/d2/example/app2'
clang -std=c99 -S -emit-llvm app2.c -o app2.ll
make[1]: Leaving directory '/home/root/d2/example/app2'

real	0m0.209s
user	0m0.123s
sys	0m0.084s
make -C app1 
make[1]: Entering directory '/home/root/d2/example/app1'
clang -std=c99 -S -emit-llvm -S -emit-llvm -g app1.c -o app1.llg
make[1]: Leaving directory '/home/root/d2/example/app1'
make -C app2 
make[1]: Entering directory '/home/root/d2/example/app2'
clang -std=c99 -S -emit-llvm -S -emit-llvm -g app2.c -o app2.llg
make[1]: Leaving directory '/home/root/d2/example/app2'

real	0m0.122s
user	0m0.083s
sys	0m0.038s
copy files to /home/root/d2/output
/home/root/d2/bin/d2: line 64: cpio: command not found
create ll2c db
llg2cdb ./app1/app1.llg
Traceback (most recent call last):
  File "/home/root/d2/bin/llg2cdb.py", line 77, in <module>
    llg2cdb(f)
  File "/home/root/d2/bin/llg2cdb.py", line 9, in llg2cdb
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app1/app1.llg'
llg2cdb ./app2/app2.llg
Traceback (most recent call last):
  File "/home/root/d2/bin/llg2cdb.py", line 77, in <module>
    llg2cdb(f)
  File "/home/root/d2/bin/llg2cdb.py", line 9, in llg2cdb
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app2/app2.llg'

real	0m0.069s
user	0m0.062s
sys	0m0.009s
create loop files
opt ./app1/app1.ll -passes='print<loops>' -disable-output > ./app1/app1.loop
opt: ./app1/app1.ll: error: Could not open input file: No such file or directory
opt ./app2/app2.ll -passes='print<loops>' -disable-output > ./app2/app2.loop
opt: ./app2/app2.ll: error: Could not open input file: No such file or directory

real	0m0.049s
user	0m0.024s
sys	0m0.024s
create dot files
ll2dot ./app1/app1.ll > ./app1/app1.dot
Traceback (most recent call last):
  File "/home/root/d2/bin/ll2dot.py", line 116, in <module>
    ll2dot(f)
  File "/home/root/d2/bin/ll2dot.py", line 29, in ll2dot
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app1/app1.ll'
ll2dot ./app2/app2.ll > ./app2/app2.dot
Traceback (most recent call last):
  File "/home/root/d2/bin/ll2dot.py", line 116, in <module>
    ll2dot(f)
  File "/home/root/d2/bin/ll2dot.py", line 29, in ll2dot
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app2/app2.ll'

real	0m0.075s
user	0m0.057s
sys	0m0.018s
create sb files
/home/root/d2/output
sb -f ./app1/app1.const ./app1/app1.dot > ./app1/app1.sb
sb -f ./app2/app2.const ./app2/app2.dot > ./app2/app2.sb

real	0m0.057s
user	0m0.038s
sys	0m0.018s
create bbdb
/home/root/d2/output
ll2db ./app1/app1.ll
Traceback (most recent call last):
  File "/home/root/d2/bin/ll2db.py", line 143, in <module>
    ll2db(f)
  File "/home/root/d2/bin/ll2db.py", line 31, in ll2db
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app1/app1.ll'
ll2db ./app2/app2.ll
Traceback (most recent call last):
  File "/home/root/d2/bin/ll2db.py", line 143, in <module>
    ll2db(f)
  File "/home/root/d2/bin/ll2db.py", line 31, in ll2db
    with open(filename, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app2/app2.ll'

real	0m0.084s
user	0m0.072s
sys	0m0.010s
create loopdb
loop2db ./app1/app1.loop
loop2db ./app2/app2.loop

real	0m0.048s
user	0m0.029s
sys	0m0.016s
create bbsbdb
sb2db ./app1/app1.sb
sb2db ./app2/app2.sb

real	0m0.047s
user	0m0.042s
sys	0m0.005s
-rw-r--r-- 1 root root 0 Mar  2 23:20 /home/root/d2/output/bbcdb
-rw-r--r-- 1 root root 0 Mar  2 23:20 /home/root/d2/output/bbdb
-rw-r--r-- 1 root root 0 Mar  2 23:20 /home/root/d2/output/bbsbdb
-rw-r--r-- 1 root root 0 Mar  2 23:20 /home/root/d2/output/loopdb
number bb 0
number bb in loops 0
rank SB

real	0m0.024s
user	0m0.020s
sys	0m0.005s
separate ll into bb files
Traceback (most recent call last):
  File "/home/root/d2/bin/ll2bb.py", line 54, in <module>
    ll2bb(f)
  File "/home/root/d2/bin/ll2bb.py", line 9, in ll2bb
    with open(file, "r") as f:
FileNotFoundError: [Errno 2] No such file or directory: './app1/app1.ll'

real	0m0.039s
user	0m0.037s
sys	0m0.001s
create normalized file for each bb file

real	0m0.001s
user	0m0.001s
sys	0m0.000s
rank order normalized files

real	0m0.001s
user	0m0.001s
sys	0m0.000s
bbselect

real	0m0.044s
user	0m0.033s
sys	0m0.010s
ID SB from BB

real	0m0.001s
user	0m0.000s
sys	0m0.003s
colorize BB within SB CFG

real	0m0.001s
user	0m0.001s
sys	0m0.000s
```

**Comment by dtkerns** [comment]
I'll check into this.

**Comment by dtkerns** [comment]
This error concerns me:
```
/home/root/d2/bin/d2: line 14: [: 14.0.0: integer expression expected
```
That indicates an issue reading the .config file created when you run `make` in the `d2/src`
I made a small change to the d2 script, to be more explicit about where to run `make`
Can you verify (show me here) the contents of your `.config` file
Also, it looks like I make a bad assumption about the availability of `cpio`. I'll list and check for additional dependencies.

**Comment by manuel-g-castro** [comment]
Sure! Here is my `./bin/.config`:

```
root@6ce3f6fc9c55:/home/root/d2/bin# cat .config 
LLVMVER=14.0.0
NEWPASSMANAGER=implemented
```

**Comment by dtkerns** [comment]
I just pushed some fixes, should print a warning about LLVM version < 19. please pull, and re-try. thanks

**Comment by manuel-g-castro** [comment]
Hey, David! Thanks for the changes.

I have recreated the container with the ubuntu 22.04 image, and noticed one thing that I forgot to mention: the `opt` dependency installs at `/usr/bin/opt-14`, instead of `/usr/bin/opt`, as expected by your Makefile. I had to create a symlink `/usr/bin/opt` in order to make to work. **I would suggest checking your package manager where and how is it installed and commenting it on the `readme.md`**.

With that done, as the image comes without a compiler, I only installed `clang` for the testing. I found out that the compilation of `d2` expects `g++`, which I tried to amend by executing `make CXX=/usr/bin/clang`. Unfortunately, it failed in the linking stage. Then, I am afraid that `d2`'s compilation requires GNU's compiler, **which should be in the dependency list**.

Also, which version of `python` I need to install? I made a symlink of 3.10.12 to `/usr/bin/python`, and it worked. **But I believe it should be clear at least if its 2.7 or 3+**.

With the current changes, I am on commit `4c071`, and the full installed software dependency, the example worked without problems. Although I find odd that it is complaining that I have an old version of LLVM, 14, and that I need to include `-O1` flag in both app's `CFLAGS`, which I have done. I do not think this is due to your fault. Maybe `clang`'s aptitude package is still shipping with an old, buggy, version.

Finally, with all the bold comments addressed, for my side we can close this issue.

For reference, I am including the files created by `d2` (after removing functions without I/O calls, as pointed out both in the paper and in the `step-by-step.md`). 

```
root@aa99ffb7b85f:/home/d2/output# find . -type f -exec ls -l {} \;
-rw-r--r-- 1 root root 90 Mar  3 17:02 ./app1/app1.bb/entry_rotf
-rw-r--r-- 1 root root 173 Mar  3 17:02 ./app1/app1.bb/L_rotf_16.norm
-rw-r--r-- 1 root root 173 Mar  3 17:02 ./app1/app1.bb/L_main_20.norm
-rw-r--r-- 1 root root 19 Mar  3 17:02 ./app1/app1.bb/entry_rotf.norm
-rw-r--r-- 1 root root 439 Mar  3 17:02 ./app1/app1.bb/L_rotf_21
-rw-r--r-- 1 root root 107 Mar  3 17:02 ./app1/app1.bb/L_rotf_14
-rw-r--r-- 1 root root 18 Mar  3 17:02 ./app1/app1.bb/L_rotf_3.norm
-rw-r--r-- 1 root root 46 Mar  3 17:02 ./app1/app1.bb/L_rotf_14.norm
-rw-r--r-- 1 root root 175 Mar  3 17:02 ./app1/app1.bb/L_main_6.norm
-rw-r--r-- 1 root root 172 Mar  3 17:02 ./app1/app1.bb/L_main_3
-rw-r--r-- 1 root root 193 Mar  3 17:02 ./app1/app1.bb/L_main_40.norm
-rw-r--r-- 1 root root 88 Mar  3 17:02 ./app1/app1.bb/L_main_37.norm
-rw-r--r-- 1 root root 46 Mar  3 17:02 ./app1/app1.bb/L_main_18.norm
-rw-r--r-- 1 root root 353 Mar  3 17:02 ./app1/app1.bb/L_rotf_21.norm
-rw-r--r-- 1 root root 113 Mar  3 17:02 ./app1/app1.bb/L_main_42.norm
-rw-r--r-- 1 root root 439 Mar  3 17:02 ./app1/app1.bb/L_main_8
-rw-r--r-- 1 root root 254 Mar  3 17:02 ./app1/app1.bb/L_main_40
-rw-r--r-- 1 root root 79 Mar  3 17:02 ./app1/app1.bb/L_rotf_3
-rw-r--r-- 1 root root 179 Mar  3 17:02 ./app1/app1.bb/L_main_42
-rw-r--r-- 1 root root 19 Mar  3 17:02 ./app1/app1.bb/entry_main.norm
-rw-r--r-- 1 root root 431 Mar  3 17:02 ./app1/app1.bb/L_rotf_4
-rw-r--r-- 1 root root 367 Mar  3 17:02 ./app1/app1.bb/L_main_8.norm
-rw-r--r-- 1 root root 118 Mar  3 17:02 ./app1/app1.bb/entry_main
-rw-r--r-- 1 root root 236 Mar  3 17:02 ./app1/app1.bb/L_main_6
-rw-r--r-- 1 root root 441 Mar  3 17:02 ./app1/app1.bb/L_main_25
-rw-r--r-- 1 root root 114 Mar  3 17:02 ./app1/app1.bb/L_main_3.norm
-rw-r--r-- 1 root root 354 Mar  3 17:02 ./app1/app1.bb/L_main_25.norm
-rw-r--r-- 1 root root 152 Mar  3 17:02 ./app1/app1.bb/L_main_37
-rw-r--r-- 1 root root 242 Mar  3 17:02 ./app1/app1.bb/L_main_20
-rw-r--r-- 1 root root 370 Mar  3 17:02 ./app1/app1.bb/L_rotf_4.norm
-rw-r--r-- 1 root root 239 Mar  3 17:02 ./app1/app1.bb/L_rotf_16
-rw-r--r-- 1 root root 107 Mar  3 17:02 ./app1/app1.bb/L_main_18
-rw-r--r-- 1 root root 1665 Mar  3 17:02 ./app1/app1.sb
-rw-r--r-- 1 root root 18152 Mar  3 17:01 ./app1/app1.llg
-rw-r--r-- 1 root root 1110 Mar  3 17:01 ./app1/app1.dot
-rw-r--r-- 1 root root 260 Mar  3 17:02 ./app1/app1.db.d/g0007.dot
-rw-r--r-- 1 root root 225 Mar  3 17:02 ./app1/app1.db.d/g0004.dot
-rw-r--r-- 1 root root 179 Mar  3 17:02 ./app1/app1.db.d/g0005.dot
-rw-r--r-- 1 root root 235 Mar  3 17:02 ./app1/app1.db.d/g0003.dot
-rw-r--r-- 1 root root 222 Mar  3 17:02 ./app1/app1.db.d/g0001.dot
-rw-r--r-- 1 root root 181 Mar  3 17:02 ./app1/app1.db.d/g0006.dot
-rw-r--r-- 1 root root 207 Mar  3 17:02 ./app1/app1.db.d/g0000.dot
-rw-r--r-- 1 root root 179 Mar  3 17:02 ./app1/app1.db.d/g0002.dot
-rw-r--r-- 1 root root 5841 Mar  3 17:01 ./app1/app1.ll
-rw-r--r-- 1 root root 0 Mar  3 17:01 ./app1/app1.loop
-rw-r--r-- 1 root root 4 Mar  3 17:02 ./app1/app1.const
-rw-r--r-- 1 root root 34 Mar  3 17:02 ./bbselected
-rw-r--r-- 1 root root 68 Mar  3 17:02 ./bbsel_norm
-rw-r--r-- 1 root root 1050 Mar  3 17:02 ./lobb
-rw-r--r-- 1 root root 3359 Mar  3 17:02 ./bbdb
-rw-r--r-- 1 root root 25 Mar  3 17:02 ./bbsel_norm.bb
-rw-r--r-- 1 root root 0 Mar  3 17:02 ./loopdb
-rw-r--r-- 1 root root 30 Mar  3 17:01 ./loll
-rw-r--r-- 1 root root 460 Mar  3 17:02 ./bbsbdb
-rw-r--r-- 1 root root 183 Mar  3 17:02 ./app2/app2.db.d/g0003.dot
-rw-r--r-- 1 root root 225 Mar  3 17:02 ./app2/app2.db.d/g0001.dot
-rw-r--r-- 1 root root 222 Mar  3 17:02 ./app2/app2.db.d/g0000.dot
-rw-r--r-- 1 root root 179 Mar  3 17:02 ./app2/app2.db.d/g0002.dot
-rw-r--r-- 1 root root 159 Mar  3 17:02 ./app2/app2.bb/L_main_16
-rw-r--r-- 1 root root 331 Mar  3 17:02 ./app2/app2.bb/L_rotf_6
-rw-r--r-- 1 root root 167 Mar  3 17:02 ./app2/app2.bb/entry_rotf
-rw-r--r-- 1 root root 241 Mar  3 17:02 ./app2/app2.bb/L_rotf_16.norm
-rw-r--r-- 1 root root 177 Mar  3 17:02 ./app2/app2.bb/L_main_5.norm
-rw-r--r-- 1 root root 68 Mar  3 17:02 ./app2/app2.bb/entry_rotf.norm
-rw-r--r-- 1 root root 200 Mar  3 17:02 ./app2/app2.bb/L_main_62
-rw-r--r-- 1 root root 227 Mar  3 17:02 ./app2/app2.bb/L_main_35.norm
-rw-r--r-- 1 root root 127 Mar  3 17:02 ./app2/app2.bb/L_rotf2_5
-rw-r--r-- 1 root root 385 Mar  3 17:02 ./app2/app2.bb/L_rotf_23
-rw-r--r-- 1 root root 280 Mar  3 17:02 ./app2/app2.bb/L_main_44
-rw-r--r-- 1 root root 92 Mar  3 17:02 ./app2/app2.bb/L_rotf_13.norm
-rw-r--r-- 1 root root 68 Mar  3 17:02 ./app2/app2.bb/entry_rotf2.norm
-rw-r--r-- 1 root root 237 Mar  3 17:02 ./app2/app2.bb/L_main_5
-rw-r--r-- 1 root root 443 Mar  3 17:02 ./app2/app2.bb/L_main_50
-rw-r--r-- 1 root root 194 Mar  3 17:02 ./app2/app2.bb/L_main_66.norm
-rw-r--r-- 1 root root 154 Mar  3 17:02 ./app2/app2.bb/L_rotf2_13
-rw-r--r-- 1 root root 254 Mar  3 17:02 ./app2/app2.bb/L_main_66
-rw-r--r-- 1 root root 95 Mar  3 17:02 ./app2/app2.bb/L_main_41.norm
-rw-r--r-- 1 root root 236 Mar  3 17:02 ./app2/app2.bb/L_main_3
-rw-r--r-- 1 root root 95 Mar  3 17:02 ./app2/app2.bb/L_main_16.norm
-rw-r--r-- 1 root root 276 Mar  3 17:02 ./app2/app2.bb/L_rotf_6.norm
-rw-r--r-- 1 root root 279 Mar  3 17:02 ./app2/app2.bb/L_main_19
-rw-r--r-- 1 root root 67 Mar  3 17:02 ./app2/app2.bb/L_rotf2_5.norm
-rw-r--r-- 1 root root 292 Mar  3 17:02 ./app2/app2.bb/L_main_35
-rw-r--r-- 1 root root 92 Mar  3 17:02 ./app2/app2.bb/L_rotf2_13.norm
-rw-r--r-- 1 root root 206 Mar  3 17:02 ./app2/app2.bb/L_main_44.norm
-rw-r--r-- 1 root root 292 Mar  3 17:02 ./app2/app2.bb/L_main_10
-rw-r--r-- 1 root root 331 Mar  3 17:02 ./app2/app2.bb/L_rotf2_6
-rw-r--r-- 1 root root 130 Mar  3 17:02 ./app2/app2.bb/L_main_62.norm
-rw-r--r-- 1 root root 19 Mar  3 17:02 ./app2/app2.bb/entry_main.norm
-rw-r--r-- 1 root root 180 Mar  3 17:02 ./app2/app2.bb/L_main_68
-rw-r--r-- 1 root root 118 Mar  3 17:02 ./app2/app2.bb/entry_main
-rw-r--r-- 1 root root 168 Mar  3 17:02 ./app2/app2.bb/entry_rotf2
-rw-r--r-- 1 root root 309 Mar  3 17:02 ./app2/app2.bb/L_rotf2_16.norm
-rw-r--r-- 1 root root 386 Mar  3 17:02 ./app2/app2.bb/L_main_25
-rw-r--r-- 1 root root 175 Mar  3 17:02 ./app2/app2.bb/L_main_3.norm
-rw-r--r-- 1 root root 298 Mar  3 17:02 ./app2/app2.bb/L_main_25.norm
-rw-r--r-- 1 root root 276 Mar  3 17:02 ./app2/app2.bb/L_rotf2_6.norm
-rw-r--r-- 1 root root 127 Mar  3 17:02 ./app2/app2.bb/L_rotf_5
-rw-r--r-- 1 root root 159 Mar  3 17:02 ./app2/app2.bb/L_main_41
-rw-r--r-- 1 root root 154 Mar  3 17:02 ./app2/app2.bb/L_rotf_13
-rw-r--r-- 1 root root 356 Mar  3 17:02 ./app2/app2.bb/L_main_50.norm
-rw-r--r-- 1 root root 205 Mar  3 17:02 ./app2/app2.bb/L_main_19.norm
-rw-r--r-- 1 root root 227 Mar  3 17:02 ./app2/app2.bb/L_main_10.norm
-rw-r--r-- 1 root root 442 Mar  3 17:02 ./app2/app2.bb/L_rotf2_25
-rw-r--r-- 1 root root 309 Mar  3 17:02 ./app2/app2.bb/L_rotf_16
-rw-r--r-- 1 root root 67 Mar  3 17:02 ./app2/app2.bb/L_rotf_5.norm
-rw-r--r-- 1 root root 381 Mar  3 17:02 ./app2/app2.bb/L_rotf2_16
-rw-r--r-- 1 root root 297 Mar  3 17:02 ./app2/app2.bb/L_rotf_23.norm
-rw-r--r-- 1 root root 355 Mar  3 17:02 ./app2/app2.bb/L_rotf2_25.norm
-rw-r--r-- 1 root root 114 Mar  3 17:02 ./app2/app2.bb/L_main_68.norm
-rw-r--r-- 1 root root 8697 Mar  3 17:01 ./app2/app2.ll
-rw-r--r-- 1 root root 0 Mar  3 17:01 ./app2/app2.loop
-rw-r--r-- 1 root root 1928 Mar  3 17:01 ./app2/app2.dot
-rw-r--r-- 1 root root 809 Mar  3 17:02 ./app2/app2.sb
-rw-r--r-- 1 root root 28860 Mar  3 17:01 ./app2/app2.llg
-rw-r--r-- 1 root root 14 Mar  3 17:02 ./app2/app2.const
-rw-r--r-- 1 root root 1613 Mar  3 17:01 ./bbcdb
-rw-r--r-- 1 root root 32 Mar  3 17:01 ./lollg
-rw-r--r-- 1 root root 27 Mar  3 17:02 ./sbselected
-rw-r--r-- 1 root root 1112 Mar  3 17:02 ./rankorder
-rw-r--r-- 1 root root 27 Mar  3 17:02 ./sbmarkup
-rw-r--r-- 1 root root 60 Mar  3 17:02 ./ronormfiles
```

**Comment by dtkerns** [comment]
hello, thanks for your continued review.

LLVM's installation of `opt` is going to be really tricky and distro dependent. (I did most of the original development of this package under Cent OS7, and switching to RHEL 9 caused a plethora of LLVM issues.) I made a note of the `opt-VV` issue in the readme.md, but without a suggested solution, I may just rely a a human to to correct it as needed. On RHEL 9, LLVM is on version 16 (Cent OS7 was LLVM version 3!) ... unfortunately, I don't expect either to ever update to v19, where LLVM developers are currently at.

d2 is c++ code, so it either needs to be `g++`,  `clang++` or another c++ compiler of your choosing (and definition of `CXX` in the `src/Makefile` . There is nothing in sb.cc that a vanilla c++ compiler can't handle; a C compiler, not so much. ;) I added additional checks in the src/Makefile for this too.

Finally, I specified explicitly `python3` as the required dependency; sorry, I should have known better.

If you're ok with these 3 items as I addressed them, I'll close the issue.
